### ISSUE TITLE
Compiz crashes when restoring decoration to a window

### ISSUE THREAD

**Issue by alekseyt** [compiz-reloaded/compiz#106]
Steps to reproduce:
1. Open ccsm -> Window Decoration
2. Set the following rule to Decoration Windows: !(state=maxvert & state=maxhorz)

Now all maximized windows are not decorated.

3. Maximize a window
4. Unmaximize a window

Result: compiz crashes.

Alternative steps to reproduce:
1. Open ccsm -> Window Decoration
2. Decoration Windows -> change any to !(ccsm window id) (you can pick id with picker) and press tab

Now ccsm window is not decorated.

3. Change Decoration Windows back to any

Now ccsm window is decorated again.

4. Repeat 2-3 until compiz crashes

In this scenario it's better to start from not decorated ccsm window, then change window id to any, then back to window id, then back to any, here compiz usually always crashes.

In my debugging attempt i've found that it crashes with the following stack trace (valgrind output):

```
==19839== Use of uninitialised value of size 8
==19839==    at 0x4C31153: memcpy@GLIBC_2.2.5 (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==19839==    by 0xEF46C6C: ??? (in /usr/lib/xorg/modules/drivers/swrast_dri.so)
==19839==    by 0xEF449C9: ??? (in /usr/lib/xorg/modules/drivers/swrast_dri.so)
==19839==    by 0x117C67: enableTexture (texture.c:423)
==19839==    by 0x141E37: enableFragmentOperationsAndDrawGeometry (paint.c:1115)
==19839==    by 0x142094: drawWindowTexture (paint.c:1183)
==19839==    by 0x177823D3: decorDrawWindow (decoration.c:239)
==19839==    by 0x14236B: paintWindow (paint.c:1267)
==19839==    by 0x16D6B8EC: movePaintWindow (move.c:871)
==19839==    by 0x16F73CA5: resizePaintWindow (resize.c:1469)
==19839==    by 0x13EE5C: paintOutputRegion (paint.c:412)
==19839==    by 0x13F5D2: paintOutput (paint.c:570)
==19839== 
==19839== Invalid write of size 2
==19839==    at 0x4C31153: memcpy@GLIBC_2.2.5 (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==19839==    by 0xEF46C6C: ??? (in /usr/lib/xorg/modules/drivers/swrast_dri.so)
==19839==    by 0xEF449C9: ??? (in /usr/lib/xorg/modules/drivers/swrast_dri.so)
==19839==    by 0x117C67: enableTexture (texture.c:423)
==19839==    by 0x141E37: enableFragmentOperationsAndDrawGeometry (paint.c:1115)
==19839==    by 0x142094: drawWindowTexture (paint.c:1183)
==19839==    by 0x177823D3: decorDrawWindow (decoration.c:239)
==19839==    by 0x14236B: paintWindow (paint.c:1267)
==19839==    by 0x16D6B8EC: movePaintWindow (move.c:871)
==19839==    by 0x16F73CA5: resizePaintWindow (resize.c:1469)
==19839==    by 0x13EE5C: paintOutputRegion (paint.c:412)
==19839==    by 0x13F5D2: paintOutput (paint.c:570)
==19839==  Address 0x118c93440 is not stack'd, malloc'd or (recently) free'd
```

It's this code in ``texture.c``:

```
	(*screen->bindTexImage) (screen->display->display,
				 texture->pixmap,
				 GLX_FRONT_LEFT_EXT,
				 NULL);
```

I've checked if texture was free'd before or pixmap was free'd before - as far as i can see they weren't. I've also checked reference counting on textures - seems ok to me. ``enableTexture()`` also used in many places of compiz, so i assumed it's somehow GL issue, not necessarily C++ issue.

My assumption is that decoration is using some outdated pixbufs/textures and copy from pixbuf to texture fails, or vise versa. I've made the following change: when window is undecorated, its decorations released completely, not only WindowDecoration struct, but also Decoration.

Now, i'm not really expert in compiz, decoration.c is also kind of lacking in comments, but here is patch:

```
From 8bc5a6a1f331d143e47374e23b53317cb7cd0cd6 Mon Sep 17 00:00:00 2001
From: Aleksey Tulinov <aleksey.tulinov@gmail.com>
Date: Fri, 26 Jan 2018 10:54:51 +0200
Subject: [PATCH] decoration: Fixed crash when re-enabling decoration on a
 window where decoration was previously disabled

---
 plugins/decoration.c | 15 ++++++++++++---
 1 file changed, 12 insertions(+), 3 deletions(-)

diff --git a/plugins/decoration.c b/plugins/decoration.c
index e2068926..4f09150a 100644
--- a/plugins/decoration.c
+++ b/plugins/decoration.c
@@ -367,7 +367,7 @@ decorReleaseTexture (CompScreen   *screen,
     DECOR_DISPLAY (screen->display);
 
     texture->refCount--;
-    if (texture->refCount)
+    if (texture->refCount > 0)
 	return;
 
     if (texture == dd->textures)
@@ -658,9 +658,11 @@ decorUpdateDecorations (CompScreen   *screen,
        if the decoration changed, and /then/ release
        and re-update it, but for now just releasing all */
     if (decors && *decorNum > 0)
-	decorReleaseDecorations (screen, decors, decorNum);
+    {
+    decorReleaseDecorations (screen, decors, decorNum);
     decors = NULL;
     *decorNum = 0;
+    }
 
     result = XGetWindowProperty (screen->display->display, id,
 				 decorAtom, 0L,
@@ -737,7 +739,7 @@ decorUpdateDecorations (CompScreen   *screen,
 	{
 	    unsigned int realDecorNum = i;
 	    decorReleaseDecorations (screen, decors, &realDecorNum);
-	    *decorNum = realDecorNum;
+	    *decorNum = 0;
 	    XFree (data);
 	    return NULL;
 	}
@@ -1333,8 +1335,15 @@ decorWindowUpdate (CompWindow *w,
 	memset (&emptyInput, 0, sizeof (emptyInput));
 	setWindowFrameExtents (w, &emptyInput);
 
+	/* decoration was already destroyed under if (old) */
 	dw->wd = NULL;
 
+	if (dw->decors && dw->decorNum > 0)
+	decorReleaseDecorations (w->screen, dw->decors, &dw->decorNum);
+
+	dw->decors = NULL;
+	dw->decorNum = 0;
+
 	moveDx = -oldShiftX;
 	moveDy = -oldShiftY;
 
-- 
2.15.1
```

Patch includes some other fixes i've made while debugging to make code more correct. Important part is at bottom:

```
+	if (dw->decors && dw->decorNum > 0)
+	decorReleaseDecorations (w->screen, dw->decors, &dw->decorNum);
```

If fix is ok in general, i can reformat the patch and remove "unneeded" bits or make a pull request (but i would prefer to send a patch instead of pull request).

**Comment by soreau** [comment]
First, I'd like to thank you for the detailed bug report and the time you've invested in this so far. I was not able to reproduce using the steps you provided but I think you might be hitting [emerald issue #31](https://github.com/compiz-reloaded/emerald/issues/31) if you're using X version >=1.19.6. Here's a [backtrace](http://sprunge.us/SGCf) I have reproducing the emerald problem with the affected version. However I did this under Xephyr and downgrading X to 1.19.5 or previous versions I tried still had the problem so I was not able to find a good commit to bisect. My system version of X is unaffected so this is probably why I cannot reproduce with your steps. I was hoping to find someone that is willing to bisect X to find the exact commit when this started happening because I'm unable to find a good commit yet. Your patch is interesting though I'm no expert in this area of the code. I'll try to round up some testers.

**Comment by alekseyt** [comment]
I'm using X 1.19.6 in virtualbox. I don't think it's the same issue, i can repro it on both emerald and gtk-window-decorator. But maybe #31 is close to this issue.

I've done some additional testing and, sadly, it looks like patch is fixing only steps to reproduce variant 1 (when single window decoration changes). Steps to repro variant 2 are still reproducible for me. Maybe because ccsm is updating decorations on all windows when settings are changed or something. Although i can't repro variant 1 this patch.

I'll try to take a further look into this. Could someone review the patch and say if makes sense at all or am i just introducing some side effects that make compiz to dodge the bug? Also any suggestions where to look would be appreciated.

**Comment by soreau** [comment]
Do you have any way to test on X version <1.19.6? The backtraces looks quite similar. So far as the emerald bug is concerned, I'm waiting on X bisect results to consult with X devs.

**Comment by soreau** [comment]
and does using --loose-binding option prevent the crashes?

**Comment by alekseyt** [comment]
I'm pretty sure this was reproducible for me before 2017/12/20 (when 1.19.6 was tagged), it's just now when i tried to approach this issue. I can't downgrade X normally, only if i build it from sources. But as i said, this was always reproducible for me in virtualbox, as far as i remember. I even was using compiz 0.9 before because it doesn't have this issue. Alas 0.9 source code is very different in this part, so i couldn't spot the difference yet.

Yes, --loose-binding solves the crashes, but introduces rendering artifacts.

**Comment by alekseyt** [comment]
Additional steps to reproduce (replacing all decorations):

1. gtk-window-decorator --replace
2. emerald --replace
3. gtk-window-decorator --replace
4. emerald --replace
5. Repeat 1-4

Seems to be producing similar stacktrace too.

**Comment by alekseyt** [comment]
If anyone wants to try a patch (or two), try one below. In my situations it fixes steps to repro 1 and 2. Steps 3 - helps, but not quite fixes. To apply patch to git or to 0.8.4 sources: 
```
patch -p1 <decorate.patch
```

I removed all "unimportant" bits, so changes more clearly visible.

```
diff --git a/plugins/decoration.c b/plugins/decoration.c
index e2068926..1be5645d 100644
--- a/plugins/decoration.c
+++ b/plugins/decoration.c
@@ -1283,6 +1283,12 @@ decorWindowUpdate (CompWindow *w,
                    decoration = NULL;
            }
        }
+
+       if (dw->decors && dw->decorNum > 0)
+       decorReleaseDecorations (w->screen, dw->decors, &dw->decorNum);
+
+       dw->decors = NULL;
+       dw->decorNum = 0;
     }
 
     if (!ds->dmWin || !allowDecoration)
@@ -1333,8 +1339,15 @@ decorWindowUpdate (CompWindow *w,
        memset (&emptyInput, 0, sizeof (emptyInput));
        setWindowFrameExtents (w, &emptyInput);
 
+       /* decoration was already destroyed under if (old) */
        dw->wd = NULL;
 
+       if (dw->decors && dw->decorNum > 0)
+       decorReleaseDecorations (w->screen, dw->decors, &dw->decorNum);
+
+       dw->decors = NULL;
+       dw->decorNum = 0;
+
        moveDx = -oldShiftX;
        moveDy = -oldShiftY;
```

BTW i saw hard X freezes like described in emerald issue when switching decorators like in steps to repro 3.

**Comment by alekseyt** [comment]
Alternative patch:
```
diff --git a/plugins/decoration.c b/plugins/decoration.c
index e2068926..7df16eff 100644
--- a/plugins/decoration.c
+++ b/plugins/decoration.c
@@ -754,9 +754,9 @@ decorWindowUpdateDecoration (CompWindow *w)
     DECOR_DISPLAY (w->screen->display);
     DECOR_WINDOW (w);
 
-    dw->decors = decorUpdateDecorations (w->screen, w->id,
-                                        dd->winDecorAtom,
-                                        dw->decors, &dw->decorNum);
+    // dw->decors = decorUpdateDecorations (w->screen, w->id,
+    //                                  dd->winDecorAtom,
+    //                                  dw->decors, &dw->decorNum);
 }
 
 static WindowDecoration *
```

Now steps 1 and 2 doesn't reproduce, also applications switcher works w/o crashes or freezes. Side effect: no buttons on window headers.

I think DecorWindow->decors is sweet spot, it was introduced in commit 75d99fec.

**Comment by alekseyt** [comment]
This supposedly fixes steps to repro 1 in non-hacky way (when decoration changes on a window):

```
From cdf67b0a42477698b5d30b52850110dfebcc57b7 Mon Sep 17 00:00:00 2001
From: Aleksey Tulinov <aleksey.tulinov@gmail.com>
Date: Sat, 27 Jan 2018 03:33:21 +0200
Subject: [PATCH] Dropping decors when property changed on a window and that
 window shouldn't be decorated. Extracted decoration check into
 decorWindowShouldDecorate(). Fixes #106 (partially).

---
 plugins/decoration.c | 69 ++++++++++++++++++++++++++++++++++------------------
 1 file changed, 45 insertions(+), 24 deletions(-)

diff --git a/plugins/decoration.c b/plugins/decoration.c
index e2068926..c3a01313 100644
--- a/plugins/decoration.c
+++ b/plugins/decoration.c
@@ -635,6 +635,42 @@ decorReleaseDecorations (CompScreen   *screen,
     }
 }
 
+static Bool
+decorWindowShouldDecorate(CompWindow *w)
+{
+    Bool	     decorate = FALSE;
+    CompMatch	     *match;
+
+    DECOR_DISPLAY (w->screen->display);
+
+    switch (w->type) {
+    case CompWindowTypeDialogMask:
+    case CompWindowTypeModalDialogMask:
+    case CompWindowTypeUtilMask:
+    case CompWindowTypeMenuMask:
+    case CompWindowTypeNormalMask:
+	if (w->mwmDecor & (MwmDecorAll | MwmDecorTitle))
+	    decorate = TRUE;
+    default:
+	break;
+    }
+
+    if (w->wmType & (CompWindowTypeDockMask | CompWindowTypeDesktopMask))
+	decorate = FALSE;
+
+    if (w->attrib.override_redirect)
+	decorate = FALSE;
+
+    if (decorate)
+    {
+	match = &dd->opt[DECOR_DISPLAY_OPTION_DECOR_MATCH].value.match;
+	if (!matchEval (match, w))
+	    decorate = FALSE;
+    }
+
+    return decorate;
+}
+
 static Decoration **
 decorUpdateDecorations (CompScreen   *screen,
 			Window       id,
@@ -754,6 +790,13 @@ decorWindowUpdateDecoration (CompWindow *w)
     DECOR_DISPLAY (w->screen->display);
     DECOR_WINDOW (w);
 
+    if (!decorWindowShouldDecorate(w))
+    {
+	if (dw->decors && dw->decorNum > 0)
+	decorReleaseDecorations (w->screen, dw->decors, &dw->decorNum);
+	dw->decors = NULL;
+    }
+    else
     dw->decors = decorUpdateDecorations (w->screen, w->id,
 					 dd->winDecorAtom,
 					 dw->decors, &dw->decorNum);
@@ -1222,30 +1265,7 @@ decorWindowUpdate (CompWindow *w,
     wd = dw->wd;
     old = (wd) ? wd->decor : NULL;
 
-    switch (w->type) {
-    case CompWindowTypeDialogMask:
-    case CompWindowTypeModalDialogMask:
-    case CompWindowTypeUtilMask:
-    case CompWindowTypeMenuMask:
-    case CompWindowTypeNormalMask:
-	if (w->mwmDecor & (MwmDecorAll | MwmDecorTitle))
-	    decorate = TRUE;
-    default:
-	break;
-    }
-
-    if (w->wmType & (CompWindowTypeDockMask | CompWindowTypeDesktopMask))
-	decorate = FALSE;
-
-    if (w->attrib.override_redirect)
-	decorate = FALSE;
-
-    if (decorate)
-    {
-	match = &dd->opt[DECOR_DISPLAY_OPTION_DECOR_MATCH].value.match;
-	if (!matchEval (match, w))
-	    decorate = FALSE;
-    }
+    decorate = decorWindowShouldDecorate (w);
 
     if (decorate)
     {
@@ -1935,6 +1955,7 @@ decorMatchPropertyChanged (CompDisplay *d,
 {
     DECOR_DISPLAY (d);
 
+    decorWindowUpdateDecoration (w);
     decorWindowUpdate (w, TRUE);
 
     UNWRAP (dd, d, matchPropertyChanged);
-- 
2.16.1
```

**Comment by alekseyt** [comment]
This supposedly fixes steps to repro 2 (when decoration match rule is changes in ccsm):

```
From 8a29584e8af4aa67931b848441e1982c157e4c98 Mon Sep 17 00:00:00 2001
From: Aleksey Tulinov <aleksey.tulinov@gmail.com>
Date: Sat, 27 Jan 2018 03:34:07 +0200
Subject: [PATCH] Updating decoration when match option changes in ccsm. Fixes
 #106.

---
 plugins/decoration.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/plugins/decoration.c b/plugins/decoration.c
index c3a01313..0648c3eb 100644
--- a/plugins/decoration.c
+++ b/plugins/decoration.c
@@ -1819,7 +1819,10 @@ decorSetDisplayOption (CompPlugin      *plugin,
 
 	    for (s = display->screens; s; s = s->next)
 		for (w = s->windows; w; w = w->next)
+		{
+		    decorWindowUpdateDecoration (w);
 		    decorWindowUpdate (w, TRUE);
+		}
 	}
 	break;
     default:
-- 
2.16.1
```

Requires previous patch to be applied too.

**Comment by alekseyt** [comment]
Regarding steps to repro 3 (decorator changes), well, it's a crash alright, but i wasn't able to fix it yet. I believe two previous patches fix originally reported issue, now decorations can be turned on and off on windows w/o crash, at least as far as i can see. But this issue can be kept open if needed, or i can submit another issue, whatever works.

Sadly, it doesn't fix compiz-reloaded/emerald#31. I think it's similar issue, maybe also related to decorations, but i'm guessing it's somewhere closer to ``switch.c``. Consider the following: shift switcher works, ring switcher works, stack switcher works, i'm seeing crashes only on application switcher and static application switcher. I would assume there is something wrong with latter two switchers but not with emerald.
